software/gpsd/Dockerfile.template

```diff
@@ -20,16 +20,56 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 
-FROM balenalib/%%BALENA_MACHINE_NAME%%-python:3.9
+FROM balenalib/%%BALENA_MACHINE_NAME%%-python:3.9-build AS build
 
-RUN install_packages gpsd gpsd-clients dbus build-essential libffi-dev python-dev libssl-dev
+# cargo for `pip install cryptography`, a transitive dependency of balena-sdk
+RUN install_packages \
+    gpsd \
+    gpsd-clients \
+    dbus \
+    cargo
+
+# Set up and activate the virtualenv
+RUN python -m venv /usr/venv
+ENV PATH="/usr/venv/bin:$PATH" VIRTUAL_ENV=/usr/venv
+
+# Prepare our Python toolchain
 RUN pip install wheel
+RUN pip install pip --upgrade
+
+# Set our working directory
+WORKDIR /usr/src/
+# Copy in just enough to install dependencies
+COPY pyproject.toml poetry.lock ./
+# Install Python packages; dependencies from pyproject.toml via poetry-core / PEP 517
+# Create an empty co2.py so that pip is not confused
```

Comment:
I guess this comment should say `gpsd.py` so _I'm_ not confused :)

Reply:
Heh, that's what I get for copy-pasting aggressively